core/modules/wdb/ChunkResource.cc

```diff
@@ -235,8 +279,112 @@ class Backend {
         }
     }
 
+    /// Run the 'query'. If it fails, terminate the program.
+    void _execLockSql(std::string const& query) {
+        LOGF_DEBUG("execLockSql %1%" % query);
+        sql::SqlErrorObject err;
+        if(!_sqlConn.runQuery(query, err)) {
+            LOGF_ERROR("Lock failed, exiting. query=%1%  err=%2%" % query % err.printErrMsg());
+            exit(EXIT_FAILURE);
+        }
+    }
+
+    /// Return the status of the lock on the in memory tables.
+    LockStatus _memLockStatus() {
+        std::string sql = "SELECT uid FROM " + _lockDbTbl + " WHERE keyId = 1;";
+        sql::SqlResults results;
+        sql::SqlErrorObject err;
+        if (!_sqlConn.runQuery(sql, results, err)) {
+            // Assuming UNLOCKED should be safe as either it must be LOCKED_OURS to continue
+            // or we are about to try to lock. Failure to lock will cause the program to exit.
+            std::ostringstream emsg;
+            emsg << "memLockStatus query failed, assuming UNLOCKED. " << sql << " err=" << err.printErrMsg();
+            LOGF_WARN("%1%" % emsg.str());
+            return UNLOCKED;
+        }
+        std::string uidStr;
+        if (!results.extractFirstValue(uidStr, err)) {
+            std::ostringstream emsg;
+            emsg << "memLockStatus unexpected results, assuming LOCKED_OTHER. err=" << err.printErrMsg();
+            LOGF_WARN("%1%" % emsg.str());
+            return LOCKED_OTHER;
+        }
+        int uid = atoi(uidStr.c_str());
+        if (uid != _uid) {
+            std::ostringstream emsg;
+            emsg << "memLockStatus LOCKED_OTHER wrong uid. Expected " << _uid << " got " << uid
+                << " err=" << err.printErrMsg();
+            LOGF_WARN("%1%" % emsg.str());
+            return LOCKED_OTHER;
+        }
+        return LOCKED_OURS;
+    }
+
+    /// Take the lock on the in memory tables, even if held by another.
+    // Terminate the program if we can't take the lock.
+    // This must be run before any other operations on in memory tables.
+    void _memLockLock() {
```

Comment:
I think `_memLockAcquire` would be a better name for this.

More importantly, I think these semantics are problematic, essentially because what you've implemented isn't a lock. You take ownership of a mysql instance immediately rather than waiting for the owner to release it, and while you do make sure that a previous worker commits suicide as soon as it notices a loss of ownership, it will not realize it immediately. Therefore it seems to me that there is some window where two workers could conceivably operate on each-others state and thus potentially return corrupted results.

To me, it makes more sense for a worker to kill itself when it notices that the mysql instance it is trying to acquire has already been claimed. This is slightly more inconvenient from an administrative standpoint - you have to take some action if a worker dies without having released its mysql instance, but I think it's a safer design.

It also seems to give slightly simpler code - acquiring a mysql instance could just be:
```
CREATE DATABASE IF NOT EXISTS q_memoryLockDb;
CREATE TABLE IF NOT EXISTS q_memoryLockDb.memoryLockTbl (whatever INT UNIQUE) ENGINE=MEMORY;
INSERT INTO q_memoryLockDb.memoryLockTbl VALUE (0);
```
(where you exit if the insert fails because someone else has already inserted 0), while explicitly releasing it would have the same effect as restarting mysql:
```
TRUNCATE TABLE q_memoryLockDb.memoryLockTbl
```
What do you think?